MELA/interface/Mela.h

```diff
@@ -347,6 +354,98 @@ class Mela{
   double selfDAZffcoupl[SIZE_AZff][2];
   // That is a lot of them!
 
+  //madMELA couplings
```

Comment:
Are you sure this is the best way to do this? This would entail creating multiple different abstractions for a set of couplings that is not defined via a list and enumerations.

For example - looking at `MadMela.h` - one would have to create arrays for various warsaw operators (in some order), then create a structure to parse that array and link them to the original values in `params_r`, which, at the end of the day, is the same thing as a reference. Also, one would need the exact same number of enumerations as there are operators listed here in this block - rather than one space being cluttered - the user is forced to ping pong back and forth between enumerations, an array, and a common block.

I understand trying to keep things consistent, but I think the benefits of a reference outweigh the negatives - otherwise it adds many lines and complications for something that would effectively be done the same via a block of references.

Reply:
> Are you sure this is the best way to do this? This would entail creating multiple different abstractions for a set of couplings that is not defined via a list and enumerations.
> 
> For example - looking at `MadMela.h` - one would have to create arrays for various warsaw operators (in some order), then create a structure to parse that array and link them to the original values in `params_r`, which, at the end of the day, is the same thing as a reference. Also, one would need the exact same number of enumerations as there are operators listed here in this block - rather than one space being cluttered - the user is forced to ping pong back and forth between enumerations, an array, and a common block.

Well, you can think of the self-defined spin zero couplings as an example. In those cases, we would be able to set `selfDHzzcoupl[resonance_index][coupling_enumerator][real_or_imaginary]=value'. What I am suggesting is a similar structure. You could communicate this structure for all couplings in MG.

> 
> I understand trying to keep things consistent, but I think the benefits of a reference outweigh the negatives - otherwise it adds many lines and complications for something that would effectively be done the same via a block of references.

Reply:
> Well, you can think of the self-defined spin zero couplings as an example. In those cases, we would be able to set `selfDHzzcoupl[resonance_index][coupling_enumerator][real_or_imaginary]=value'. What I am suggesting is a similar structure. You could communicate this structure for all couplings in MG.

This still feels like too much - especially given that all of these EFT terms are real valued. It feels much less natural for someone to say `Mela.mdl_ch = 1` than it does `Mela.MadgraphCouplings[gHiggs_ch] = 1`, because it's an unnecessary step in the process. 

Rather than the user assigning a value directly to the reference, we now have to:
1. create and instantiate an array to be passed with X coupling indices
2. create a map with X values between the enum and a reference to the common block to link the array and the MG operator values
3. create a function to assign each index to its reference

This seems very redundant for something that, in the end, will use the exact same reference procedure, just with more work and obfuscation.

Reply:
> > Well, you can think of the self-defined spin zero couplings as an example. In those cases, we would be able to set `selfDHzzcoupl[resonance_index][coupling_enumerator][real_or_imaginary]=value'. What I am suggesting is a similar structure. You could communicate this structure for all couplings in MG.
> 
> This still feels like too much - especially given that all of these EFT terms are real valued. It feels much more natural for someone to say `Mela.mdl_ch = 1` than it does `Mela.MadgraphCouplings[gHiggs_ch] = 1`, because it's an unnecessary step in the process.

While I understand where you are coming from, i.e., the new Python implementation, the C++ interface is the primary one, and it needs to be well-maintainable. In that sense, adding coupling references to Mela is something to avoid. I should also note that the Python interface is not the only interface to Mela either, so how the different interfaces make it flexible to access the couplings from the original C++ implementation is a secondary matter of concern for the design of the C++ interface itself.

The fact that EFT terms are real-valued is also not a major concern - the struct could easily assume real-valued couplings (you just do not add the second index [2] in the array definition).

> 
> Rather than the user assigning a value directly to the reference, we now have to:
> 
> 1. create and instantiate an array to be passed with X coupling indices

That's just a matter of writing standard default and copy constructors for a coupling set, see for example https://github.com/JHUGen/JHUGenMELA/blob/master/MELA/src/TCouplings.cc#L12-L14.

> 2. create a map with X values between the enum and a reference to the common block to link the array and the MG operator values
> 3. create a function to assign each index to its reference

This is something you have to have done (in this case, in linking the references in Mela and those in the Fortran common block), so one does not need to write extra code for it any longer, aside from changing the location of the assignment in the code, and perhaps minor renaming as necessary. See for example https://github.com/JHUGen/JHUGenMELA/blob/master/MELA/src/TUtil.cc#L3360 for MCFM couplings.

References to the common block are not needed. As shown in the example, linking the Fortran library automatically makes the couplings available, in the format that is visible when executing `nm [library.so]`.

Please also note that if you have any booleans or plan to have them in the future, you need to be careful about how many bytes they end up consuming in the original Fortran implementation. For example, `noglue` is defined as a `bool` for MCFM: https://github.com/JHUGen/JHUGenMELA/blob/master/MELA/interface/TMCFM.hh#L269, but in this JHUGen function, https://github.com/JHUGen/JHUGenMELA/blob/master/MELA/interface/TModJHUGenMELA.hh#L24, `usewwcoupl` is defined to have the type `int` even though the Fortran subroutine defines it as a `logical`, https://github.com/JHUGen/JHUGenMELA/blob/master/MELA/fortran/mod_JHUGenMELA.F90#L130. Misusing the types could cause memory offsets that are not easy to track down (because they will look ok *most of the time*).

> 
> This seems very redundant for something that, in the end, will use the exact same reference procedure, just with more work and obfuscation.

Reply:
> While I understand where you are coming from, i.e., the new Python implementation, the C++ interface is the primary one, and it needs to be well-maintainable. In that sense, adding coupling references to Mela is something to avoid. I should also note that the Python interface is not the only interface to Mela either, so how the different interfaces make it flexible to access the couplings from the original C++ implementation is a secondary matter of concern for the design of the C++ interface itself.

The implementation of the Python bindings are agnostic to the way that the C++ couplings are done - they would be there regardless of how the C++ is implemented. In fact, it would probably be easier for the Python bindings to operate via the arrays since those are tried and tested. My issues lie with a perceived incongruity between the stated design philosophy and the adherence to previous structure when (I believe it to be) unnecessary.

> The fact that EFT terms are real-valued is also not a major concern - the struct could easily assume real-valued couplings (you just do not add the second index [2] in the array definition).

Yes fair enough - but my point here was more to state that a new structure is unnecessary, imo.
 
> That's just a matter of writing standard default and copy constructors for a coupling set, see for example https://github.com/JHUGen/JHUGenMELA/blob/master/MELA/src/TCouplings.cc#L12-L14.

I think we're getting closer to something here. My issue does not lie with any interface, but rather that I believe the nature of the EFT terms do not lend themselves to an array - and I think that making a new object solely for the purposes of assigning terms is unnecessary. In vein with the design philosophies you were saying earlier, I think it makes for very redundant code!

> > 2. create a map with X values between the enum and a reference to the common block to link the array and the MG operator values
> > 3. create a function to assign each index to its reference
> 
> This is something you have to have done (in this case, in linking the references in Mela and those in the Fortran common block), so one does not need to write extra code for it any longer, aside from changing the location of the assignment in the code, and perhaps minor renaming as necessary. See for example https://github.com/JHUGen/JHUGenMELA/blob/master/MELA/src/TUtil.cc#L3360 for MCFM couplings.

Except in this case, I would have to here, right? I would have to generate a new array in the SpinZeroCouplings class, then write a SpinZeroCouplings method to assign each array element (let's call it `SetEFTCouplings`), and write a bunch of enums to do so. It's a bunch of ping ponging around that involves writing much more code that is not self-explanatory (i.e. passing an array of enumerated size to a function in a class that then takes the enumerations and assigns terms to global variables), rather than simply making a block of reference assignments (that I am more than happy to move to the source file).

> References to the common block are not needed. As shown in the example, linking the Fortran library automatically makes the couplings available, in the format that is visible when executing `nm [library.so]`.

When I say "reference the common block" I mean reference the variables in the extern "C" block within the header file. I understand that they are available upon compilation from the original .so file. What I meant here was that every index in any array would need to be manually linked to the global variables in the extern block anyways - so why obfuscate where it's done?

> Please also note that if you have any booleans or plan to have them in the future, you need to be careful about how many bytes they end up consuming in the original Fortran implementation. For example, `noglue` is defined as a `bool` for MCFM: https://github.com/JHUGen/JHUGenMELA/blob/master/MELA/interface/TMCFM.hh#L269, but in this JHUGen function, https://github.com/JHUGen/JHUGenMELA/blob/master/MELA/interface/TModJHUGenMELA.hh#L24, `usewwcoupl` is defined to have the type `int` even though the Fortran subroutine defines it as a `logical`, https://github.com/JHUGen/JHUGenMELA/blob/master/MELA/fortran/mod_JHUGenMELA.F90#L130. Misusing the types could cause memory offsets that are not easy to track down (because they will look ok _most of the time_).

Yes - As I understand from [here](http://www.yolinux.com/TUTORIALS/LinuxTutorialMixingFortranAndC.html), this is true - and good advice!

Reply:
Hmm, let me also approach from a different direction then: Is there a non-degeneracy between the couplings already in JHUGen and those that are introduced for MG? Given that you mention these are spin-0 couplings, can't we simply translate JHUGen/MCFM couplings, like how it is done in the Lexicon? In fact, it could be easier to do here than in Lexicon, given that we already have access to all of the couplings/masses from extern bindings.

> It's a bunch of ping ponging around that involves writing much more code that is not self-explanatory (i.e. passing an array of enumerated size to a function in a class that then takes the enumerations and assigns terms to global variables), rather than simply making a block of reference assignments (that I am more than happy to move to the source file).

The problem with this approach would be later in implementing parallelization. We will for sure have to figure out how to handle the Fortran component either, since we would not want a serialization bottleneck right in the middle of the ME calculation workflow, but another serialization bottleneck would be introduced if we do not compartmentalize users' specification of couplings and the assignment of the values of extern-referenced couplings in the ME libraries. The current implementation was designed to be able to prevent this secondary bottleneck in the future.

Let me give you also the model that was ultimately in mind in the design: One could think of a parallelization model where we have a collection of `C(Nevents, NMEs, Ncouplings)` couplings (in ordinary event processing, `C(:, :, k)` have the same values for a given `k`, but different use cases do not have to keep the same couplings), and `I(Nevents, NMEs, Nreco)` input candidates. Then, one could write a workflow that computes the collection of MEs `F(Nevents, NMEs, Nreco)[C, I]`. Currently, the coupling structures allow us to collect `C(Nevents, NMEs, Ncouplings)` in a structured way as `S(Nevents, NMEs)[C]`, allowing us to write `F(Nevents, NMEs, Nreco)[S[C], I]`. Computations over the rest of the indices happen serially as the user executes the ME calls at the moment, but given that we have the structure to collect many input candidates, we could add the feature for batch computation, either over the flattened vector of inputs `I` as it is currently implemented (as a 1D `vector`), or still in series for `(Nevents, NMEs)`, as the user would like.

What you propose breaks the structuring `S[C]` explicitly, introducing a further complication down the road, and I would prefer to avoid that. It is not just a consistency matter with the implementation of the rest of couplings; this is the reason why we preserved coupling structures even though for serial processing, it might seem like introducing an unnecessary layer of complication.

Does what I propose require implementation of batch processing capabilities? Yes, but the current implementation reduces that to a matter of passing batches of couplings and ME specifications, and returning a batch of numbers at the end of the call. Internally, we would have to figure out how to parallelize the Fortran code, which is where I would suggest to really focus our energy.

Reply:
> Hmm, let me also approach from a different direction then: Is there a non-degeneracy between the couplings already in JHUGen and those that are introduced for MG? Given that you mention these are spin-0 couplings, can't we simply translate JHUGen/MCFM couplings, like how it is done in the Lexicon? In fact, it could be easier to do here than in Lexicon, given that we already have access to all of the couplings/masses from extern bindings.

Currently I would hold off on the Lexicon translations - Lexicon doesn't cover every operator here - and I don't really expect it to anytime soon.

> > It's a bunch of ping ponging around that involves writing much more code that is not self-explanatory (i.e. passing an array of enumerated size to a function in a class that then takes the enumerations and assigns terms to global variables), rather than simply making a block of reference assignments (that I am more than happy to move to the source file).
> 
> The problem with this approach would be later in implementing parallelization. We will for sure have to figure out how to handle the Fortran component either, since we would not want a serialization bottleneck right in the middle of the ME calculation workflow, but another serialization bottleneck would be introduced if we do not compartmentalize users' specification of couplings and the assignment of the values of extern-referenced couplings in the ME libraries. The current implementation was designed to be able to prevent this secondary bottleneck in the future.
> 
> Let me give you also the model that was ultimately in mind in the design: One could think of a parallelization model where we have a collection of `C(Nevents, NMEs, Ncouplings)` couplings (in ordinary event processing, `C(:, :, k)` have the same values for a given `k`, but different use cases do not have to keep the same couplings), and `I(Nevents, NMEs, Nreco)` input candidates. Then, one could write a workflow that computes the collection of MEs `F(Nevents, NMEs, Nreco)[C, I]`. Currently, the coupling structures allow us to collect `C(Nevents, NMEs, Ncouplings)` in a structured way as `S(Nevents, NMEs)[C]`, allowing us to write `F(Nevents, NMEs, Nreco)[S[C], I]`. Computations over the rest of the indices happen serially as the user executes the ME calls at the moment, but given that we have the structure to collect many input candidates, we could add the feature for batch computation, either over the flattened vector of inputs `I` as it is currently implemented (as a 1D `vector`), or still in series for `(Nevents, NMEs)`, as the user would like.
> 
> What you propose breaks the structuring `S[C]` explicitly, introducing a further complication down the road, and I would prefer to avoid that. It is not just a consistency matter with the implementation of the rest of couplings; this is the reason why we preserved coupling structures even though for serial processing, it might seem like introducing an unnecessary layer of complication.
> 
> Does what I propose require implementation of batch processing capabilities? Yes, but the current implementation reduces that to a matter of passing batches of couplings and ME specifications, and returning a batch of numbers at the end of the call. Internally, we would have to figure out how to parallelize the Fortran code, which is where I would suggest to really focus our energy.

Ah okay - now it makes much more sense why we do this. I think it would be best if I went down the SpinZeroCouplings route for now. In the future (AKA once we have the madMELA fleshed out for more processes + maybe offshell, etc.) we can start working to implement Lexicon's language into MELA itself - that seems like a good task.

I'm glad we worked this through!